### ISSUE TITLE
`Remove all the unused imports` does not recognize the trait method used by `write!`

### ISSUE THREAD

**Issue by A4-Tacks** [rust-lang/rust-analyzer#17729]
**rust-analyzer version**: rust-analyzer 1.80.0 (05147895 2024-07-21)

**rustc version**: rustc 1.80.0 (051478957 2024-07-21)

**editor or extension**: Vim 9.1 coc-rust-analyzer

**code snippet to reproduce**: 
```rust
use std::fmt::Write;

fn main() {
    let mut s = String::new();
    write!(s, "abc").unwrap();
}
```
**on line 1 use `Remove all the unused imports`**:
```rust

fn main() {
    let mut s = String::new();
    write!(s, "abc").unwrap();
}
```

**Comment by ShoyuVanilla** [comment]
MCVE:
```rust
//- /foo.rs
pub struct Foo {}

pub trait Bar {
    fn bar(&self) {}
}

impl Bar for Foo {}

//- /main.rs
mod foo;

use foo::Bar$0; // Trigger assist here

macro_rules! needs_bar {
    ($dst:expr) => {
        $dst.bar()
    };
}

fn main() {
    let x = foo::Foo {};
    needs_bar!(x);
}
```

**Comment by ChayimFriedman2** [comment]
That is because this assist is based on find usages for imports, which works textually, which fundamentally cannot find usages within macros.

Besides fixing the fundamental issue (find usages in macro), which will be very hard and perhaps not worth it, the way to fix that is to instead have a diagnostics for unused imports that is reported during lowering (and therefore can reliably detect whether an import is used), and have its quickfix removes the import. This will require some trickery to coordinate remove *all* imports, though.
